Add data information file and update some crontab scripts

PaulWessel opened the pull request:
This PR  makes a few necessary changes to the administration scripts and setup in anticipation of new data sets and data reorganization as discussed in #37:

1. It places information about data sets in a new file **gmt_data_server.txt** in the new information directory. This file will be synced to users, just like the hash table, and it will allow us to add new data without requiring a GMT version update.  It will replace the current hardwired information in gmt_datasets.h.  It cannot be merged with the hash table since GMT <= 6.0.0 still needs that simpler format and file.
2. The **srv_git_update.sh** script now places any updated gmt_data_server.txt file in the data dir as needed
3. I abstracted away the name of the hash file in **srv_update_sha256.sh** via the variable GMT_HASH_TABLE since I can then run this manually and experiment with new tables without breaking the system for other users.

seisman commented on scripts/srv_update_sha256.sh:
```diff
@@ -6,6 +6,9 @@
 # If an error is detected we quit.
 
+# Name of hash table
+GMT_HASH_TABLE="gmt_hash_server"
```
Are we going to use `gmt_hash_server.txt` or `gmt_data_server.txt`?

PaulWessel replied:
No change, gmt_hash_server.txt since GMT 5/6.0 knows of no other name

PaulWessel replied:
The above is about a new file that removes the hardwired information from gmt_datasets.h and gmt_remote.h into a live file.

seisman replied:
Is `gmt_data_server.txt` is manually updated? What about the data in the cache directory?

PaulWessel replied:
gmt_data_server.txt is manually updated, just like the large data grids themselves.  They are not in git because of huge size.  The gmt_data_server.txt file will be in git of course and we will make new entries for new datasets we add.  I believe this is a relatively rare task.
The data in the cache directory is not really supported data sets, they are a hodge-podge of files we need for various examples and tests, and it makes it possible for others to run those examples and tests as well.  There is no guarantee taht any files in cache will be supported.  If anyone is using trench.txt from cache it may go away one day.

seisman replied:
Sorry, I'm still confused. 

We're going to have two hash tables in the data server, the old `gmt_hash_server.txt` and the new "gmt_data_server.txt". GMT<=6.0.0 only read the `gmt_hash_server.txt` file, and GMT>=6.1 only read the `gmt_data_server.txt`, right? If that's true, how does GMT6.1 know if a cache file is available?

PaulWessel replied:
Here is the deal:

1. The gmt_hash_server.txt remains exactly as is.  Its job remains to inform GMT if a file should be updated (because there is a newer of of different size on the server) or if it should even be removed (because it is no longer on the server). All GMT version > 5.? knows about this file and its purpose and specific format remains the same.
2. GMT 6.1 will _in addition_ look for the gmt_data_server.txt file.  Its purpose is to provide information that we report the first time a file is downloaded (the attribution) plus it tells GMT what paths to build for the curl call.  The old system only knows about earth_relief and had a single directory for everything, except cache and srtm1|3.  Now, we may suddenly add a new dir earth_age and a bunch of files.  That info will go into the gmt_data_server.txt manually, and the size and hash info gets updated in the gmt_hash_server.txt automatically.

Is this clearer?

seisman replied:
Yes, good explanations. Now it sounds reasonable to me.

seisman commented on information/gmt_data_server.txt:
```diff
@@ -0,0 +1,59 @@
+# Information about all the remote data sets available on the GMT Data server
```
If I understand it correctly, gmt_remote.c in GMT>=6.1 will read this file, then get the directory name, file name and other information. If that's true, we should use `server/earth_relief` instead of `data/server/earth_relief`, because the `data` directory is invisible to users.

PaulWessel replied:
We could do that.  I put data in there so the file did not have an empty column for the symbolic links.  Suggestion? Put a period there?

seisman replied:
A period is OK. Or we could use `/` instead. 

`/` is the "root" of the data server, `/      earth_relief_01d.grd` will be mapped to https://oceania.generic-mapping-tools.org/earth_relief_01.grd.

PaulWessel replied:
OK, I like the root /.

PaulWessel commented:
An alternative is to remove the symbolic links from gmt_data_server.txt and just deal with that as a special case in the code.

seisman commented on information/gmt_data_server.txt:
```diff
@@ -0,0 +1,59 @@
+#
+# Special names for files based on tiles.  These are virtual files that never exist
+#
+/srtm3	earth_relief_03s.grd	03s	p	6.8G	Earth Relief at 3x3 arc seconds tiles provided by SRTMGL3 (land only) [NASA/USGS]
+/srtm1	earth_relief_01s.grd	01s	p	 41G	Earth Relief at 1x1 arc seconds tiles provided by SRTMGL1 (land only) [NASA/USGS]
+/srtm3	earth_relief_03s_p.grd	03s	p	6.8G	Earth Relief at 3x3 arc seconds tiles provided by SRTMGL3 (land only) [NASA/USGS]
+/srtm1	earth_relief_01s_p.grd	01s	p	 41G	Earth Relief at 1x1 arc seconds tiles provided by SRTMGL1 (land only) [NASA/USGS]
```
Only pixel registration for srtm1 and srtm3? I'm OK with it, just to confirm if they're typos.

PaulWessel replied:
Yes, those are the original files, in pixel registration only.  Cannot convert as per my earlier explanations. And because of tiling, not possible to sample a gridlines at the corners of a tile.

PaulWessel commented:
I will make one change.  The hash table starts with a line that has the number of entries.  This simplifies allocation and read, and that number is generated so not subject to human neglect.  I want the same for the gmt_data_server.txt file, so I will have the update script count the lines in the file in git, create that initial record, then append the file.  That way that number is also foolproof.

seisman commented on information/gmt_data_server.txt:
```diff
@@ -0,0 +1,59 @@
+/server/earth_relief	earth_relief_02m_p.grd	02m	p	 71M	Earth Relief at 2x2 arc minutes from Gaussian Cartesian filtering (3.7 km fullwidth) of SRTM15+V2.1 [Tozer et al., 2019]
+/server/earth_relief	earth_relief_01m_p.grd	01m	p	258M	Earth Relief at 1x1 arc minutes from Gaussian Cartesian filtering (1.9 km fullwidth) of SRTM15+V2.1 [Tozer et al., 2019]
+/server/earth_relief	earth_relief_30s_p.grd	30s	p	935M	Earth Relief at 30x30 arc seconds from Gaussian Cartesian filtering (1.0 km fullwidth) of SRTM15+V2.1 [Tozer et al., 2019]
+/server/earth_relief	earth_relief_15s_p.grd	15s	p	3.2G	Earth Relief at 15x15 arc seconds provided by SRTM15+V2.1 [Tozer et al., 2019]
```
Again, no gridline registration for 15s. Is it on purpose?

PaulWessel replied:
Yes.  The 15s pixel is the source.  We cannot derive a gridline file form that without loosing all the information that sets it apart from the 30s file.  The only real solution is to ask Dave and Brook to also create a 15s gridline file from their original sources.

PaulWessel commented:
Thanks, I will wait for server to update to make sure the new file is uploaded.  Then I will create the symbolic links and remove the old grids.  I tested this with one so it should be fine.

PaulWessel commented:
Since the first run of srv_git_update will run the old script, it will take two iterations to do something new.  So I tried to pull manually:

```
-bash-4.2$ git pull
remote: Enumerating objects: 15, done.
remote: Counting objects: 100% (15/15), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 11 (delta 6), reused 6 (delta 3), pack-reused 0
error: insufficient permission for adding an object to repository database .git/objects
fatal: failed to write object
fatal: unpack-objects failed
-bash-4.2$ ls
LICENSE  README.md  cache  recipes  scripts  staging
-bash-4.2$ ls -a
.  ..  .git  .gitignore  LICENSE  README.md  cache  recipes  scripts  staging
-bash-4.2$ ls -la
total 28
drwxrwxr-x 7 pwessel gmt  126 Feb 11 19:00 .
drwxrwxr-x 8 pwessel gmt  132 May 11 14:42 ..
drwxrwxr-x 8 pwessel gmt  198 May 22 14:06 .git
-rw-rw-r-- 1 pwessel gmt   58 Oct  4  2019 .gitignore
-rw-rw-r-- 1 pwessel gmt 7652 Sep 25  2019 LICENSE
-rw-r--r-- 1 pwessel gmt 1491 Feb 11 19:00 README.md
drwxrwxr-x 2 pwessel gmt 8192 May 19 09:27 cache
drwxrwxr-x 2 pwessel gmt   74 May 19 20:00 recipes
drwxrwxr-x 2 pwessel gmt  152 May 19 20:00 scripts
drwxrwxr-x 2 pwessel gmt   23 Oct  3  2019 staging
```

I dont understand permission issues.  The crontab runs on my (pwessel) account and I am doing this as pwessel.  Looks like I have permission, but I see in .git there is this

-rw-r--r--   1 seisman gmt    41 May 19 11:28 ORIG_HEAD

so we would need rw privileges for the group set somewhere?

PaulWessel commented:
I just did a chown for now.

PaulWessel commented:
Update is now working and the gmt_data_server.txt is now present.
I have deleted all the old grids and replaced with symbolic links to the new files.  Hash table has been updated. Please let me know if there is any odd behavior for master (or 6.0).  I am working in the branch new-remote-files.

PaulWessel commented:
Question for @seisman: So my branch is coming along and seems to work so far. Here are two issues I am working on:

1. The way I build the URL, the symbolic link names end up with https://oceania.generic-mapping-tools.org//earth_relief_01d.grd.  The double slash seems to be always ignored in macOS and Linux I think so it works.  Not sure about Windows etc.  I think it is best to have clean paths, so that means the gmt_data_server file needs to have dirs that end with slash, just like that single slash does.  So I will list dirs as "/" and "/earth_relief/" and then this problem goes away.  OK?
2. While getting files from the remote dirs works fine, what about locally?  Right now all the grids are placed (as before) in ./gmt/server, not ./gmt/server/earth_relief, for instance.  Writing there is simple since I have the directory name when I am curling.  However, when the user wants to access this file later, there is a different machinery at play, via gmt_access which will look for the file via gmt_getdatapath.  That function looks in many places, but it has no knowledge of directories beneath server.  Here we have hardwired list of dirs, including the cache dir, server, and server/srtm1|3.  If we dump all grids and images that we serve in server locally, then all will work as is.  However, I think we should match the layout on the remote server so I see no other option than to determine all the subdirs in server and loop over them instead.  This has the benefit of not having hardwired things beyond "server" and it allows us to create and find new directories driven by updates on gmtserver. To keep it simple, I will let server be special and loop over all directories found in server.  One-level only. Agree?

seisman commented:
> So I will list dirs as "/" and "/earth_relief/" and then this problem goes away.

Yes, it's even better.

seisman commented:
> I think we should match the layout on the remote server 

I agree.

> To keep it simple, I will let server be special and loop over all directories found in server. One-level only. Agree?

"loop over all directories found in server": are you talking about the local server directory or the remote GMT data server? The local server directory makes sense to me and I agree.

PaulWessel commented:
Yes, meant the local directory.  No need for loop on server since we have gmt_data_server.txt to guide us.

PaulWessel commented:
OK, I think it is all working in this branch now:

1. Old @earth_relief_xxy links works, leads to download of new data with old names and placed in server.
2. New @earth_relief_xxy_g|p works, leads to download of data into server/earth_relief

Once downloaded, @earth_relief_xyy and @earth_relief_xxy_g|p are both found locally in their different directories.  Note: No marbles listed yet in gmt_data_server.txt so not available (despite being on the server).  Need to resolve naming first.  Just because someone who already knows about BlueMarble will make sense of that name, anyone else will not know what that is, but might guess what earth_day.tif is, or earth_dayimage.tif, whatever.  It also would keep everything earth_* which I like.

See if you can break it with master or this branch, please.

The only downside I can see now is that new users who use the old names then switch to new names will get both sets of files stored and waste disk space.  Likewise, old users switching to the new names will get the new files yet the old files still take up space.  I think all we can do is to add to the documentation that because of this change they may want to exercise `gmt clear data`.

And of course, need to test this on WIndows, @joa-quim.  Please build this branch, delete your earth_relief_01d.grd file (if you have it), and then try

gmt grdinfo @earth_relief_01d -Vd
gmt grdinfo @earth_relief_01d_g -Vd
gmt grdinfo @earth_relief_01d_p -Vd

and then repeat all three.  First time it will need to download, while secod time it should fine the files locally.

joa-quim commented:
Well, it works but it also deleted all previous *earth_relief_xxx* I had
```
$ grdinfo @earth_relief_01d -Vd
grdinfo [DEBUG]: Local file c:/j/.gmt/server/gmt_hash_server.txt found
grdinfo [DEBUG]: File c:/j/.gmt/server/gmt_hash_server.txt older than 24 hours, get latest from server.
grdinfo [INFORMATION]: Downloading file https://oceania.generic-mapping-tools.org/gmt_hash_server.txt ...
grdinfo [DEBUG]: Rename c:/j/.gmt/server/gmt_hash_server.txt to c:/j/.gmt/server/gmt_hash_server.txt.old
grdinfo [DEBUG]: Rename c:/j/.gmt/server/gmt_hash_server.txt -> c:/j/.gmt/server/gmt_hash_server.txt.old
grdinfo [DEBUG]: Rename c:/j/.gmt/server/gmt_hash_server.txt.new to c:/j/.gmt/server/gmt_hash_server.txt
grdinfo [DEBUG]: Rename c:/j/.gmt/server/gmt_hash_server.txt.new -> c:/j/.gmt/server/gmt_hash_server.txt
grdinfo [DEBUG]: Load contents from c:/j/.gmt/server/gmt_hash_server.txt
grdinfo [WARNING]: File c:/j/.gmt/server/gmt_hash_server.txt said it has 226 records but only found 198 - download error???
grdinfo [WARNING]: File c:/j/.gmt/server/gmt_hash_server.txt will be deleted.  Please try again
grdinfo [DEBUG]: Load contents from c:/j/.gmt/server/gmt_hash_server.txt.old
grdinfo [DEBUG]: Look for file earth_relief_01d.grd in c:/j/.gmt
grdinfo [DEBUG]: Look for file earth_relief_01d.grd in c:/j/.gmt/cache
grdinfo [DEBUG]: Look for file earth_relief_01d.grd in c:/j/.gmt/server
grdinfo [DEBUG]: Look for file earth_relief_01m.grd in c:/j/.gmt
grdinfo [DEBUG]: Look for file earth_relief_01m.grd in c:/j/.gmt/cache
grdinfo [DEBUG]: Look for file earth_relief_01m.grd in c:/j/.gmt/server
grdinfo [DEBUG]: Found file c:/j/.gmt/server/earth_relief_01m.grd
grdinfo [DEBUG]: File earth_relief_01m.grd no longer supported on server - deleting local copy.
grdinfo [DEBUG]: Delete c:/j/.gmt/server/earth_relief_01m.grd
...
```

joa-quim commented:
and when I repeat the command it reports searching and finding the file over and over

```
grdinfo @earth_relief_01d -Vd
...
grdinfo [DEBUG]: Look for file earth_relief_01d.grd in c:/j/.gmt
grdinfo [DEBUG]: Look for file earth_relief_01d.grd in c:/j/.gmt/cache
grdinfo [DEBUG]: Look for file earth_relief_01d.grd in c:/j/.gmt/server
grdinfo [DEBUG]: Found file c:/j/.gmt/server/earth_relief_01d.grd
grdinfo [DEBUG]: Look for file earth_relief_01d.grd in c:/j/.gmt
grdinfo [DEBUG]: Look for file earth_relief_01d.grd in c:/j/.gmt/cache
grdinfo [DEBUG]: Look for file earth_relief_01d.grd in c:/j/.gmt/server
grdinfo [DEBUG]: Found file c:/j/.gmt/server/earth_relief_01d.grd
grdinfo [DEBUG]: Look for file earth_relief_01d.grd in c:/j/.gmt
grdinfo [DEBUG]: Look for file earth_relief_01d.grd in c:/j/.gmt/cache
grdinfo [DEBUG]: Look for file earth_relief_01d.grd in c:/j/.gmt/server
grdinfo [DEBUG]: Found file c:/j/.gmt/server/earth_relief_01d.grd
grdinfo [DEBUG]: Look for file earth_relief_01d.grd in c:/j/.gmt
grdinfo [DEBUG]: Look for file earth_relief_01d.grd in c:/j/.gmt/cache
grdinfo [DEBUG]: Look for file earth_relief_01d.grd in c:/j/.gmt/server
grdinfo [DEBUG]: Found file c:/j/.gmt/server/earth_relief_01d.grd
grdinfo [DEBUG]: Object ID 1 : Registered Grid File earth_relief_01d.grd as an Input resource with geometry Surface [n_objects = 2]
grdinfo [DEBUG]: gmtapi_begin_io: Input resource access is now enabled [container]
grdinfo [DEBUG]: gmtapi_import_grid: Passed ID = 1 and mode = 1
grdinfo [DEBUG]: Look for file earth_relief_01d.grd in c:/j/.gmt
grdinfo [DEBUG]: Look for file earth_relief_01d.grd in c:/j/.gmt/cache
grdinfo [DEBUG]: Look for file earth_relief_01d.grd in c:/j/.gmt/server
grdinfo [DEBUG]: Found file c:/j/.gmt/server/earth_relief_01d.grd
grdinfo [DEBUG]: Found file c:/j/.gmt/server/earth_relief_01d.grd
...
```

PaulWessel commented:
Thanks, so the deleting happens normally when we detect a file has been updated.  Since all the earth files have been updated then that is what happens.  I think we should spell that out in the dataset page so that is not a surprise at least, but this is how we wanted it to work.

As for the repeated checking: Perhaps we can try to be more clever here.  The reason it happens so many times is this:

1. In parse we call `if (gmt_check_filearg (GMT, '<', opt->arg, GMT_IN, GMT_IS_GRID))` and it ends up not finding the file and hence downloading it.  And that is just in parse.
2. Next comes` if ((G = GMT_Read_Data (API, GMT_IS_GRID, GMT_IS_FILE, GMT_IS_SURFACE, GMT_CONTAINER_ONLY, NULL, opt->arg, NULL)) == NULL)` and again we need to find the file because opt->arg has no information on the path we found earlier.
3. Finally we read with `if (GMT_Read_Data (API, GMT_IS_GRID, GMT_IS_FILE, GMT_IS_SURFACE, GMT_DATA_ONLY, wesn, opt->arg, G) == NULL) `and again we start from scratch trying to find the file.

So that was three but apparently there is more going on since you report 6 times. So even if this takes msecs it is a bit silly.  Possible changes:

1. For files given via an option, we could let the option structure also have char *path and have gmt_check_filearg be modified to report and set the full path if found.  Then later, we can use Ctrl->D.path if not NULL.
2. For command line files we could add a function that determines the path to all input files on the command line.  This array or linked structure of paths could then be used in the loop over argument instead of the opt->arg file name.

Since the actual path is found so deep in the system I think the only place where we could make a difference would be in _gmt_check_filearg_.

PaulWessel commented:
Alternatively, let gmt_check_filearg take &)opt->arg) and when path is found replace the arg with the path.  Next time it is used opt->arg is a full path and there is no searching.  Perhaps that is the slickest solution?

joa-quim commented:
> so the deleting happens normally when we detect a file has been updated.

But files were not updated (the hash maybe). They have the same name and content. But note that all grids were deleted when only one was requested.

> So even if this takes msecs it is a bit silly.

The msecs are a nothing but it makes a hell on debugging. The ``&(opt->arg)`` seems indeed the most clean solution.

PaulWessel commented:
Also, we try to be nice to the user and call gmt_access first with flag F_OK to see if the file even exists, and if it does we call gmt_access again with R_OK to see if we can read.  Each of the gmt_access calls results in the above search through directories.  With _gmt_check_filearg_ being called by all (?) modules that has input files, it seems to obvious place to do these checks once and then store the path, if the file was found.  The other not so obvious thing is that MB-System uses _gmt_check_filearg_ so we would need to call a new function _gmt_check_and_set_path_ () instead that calls _gmt_check_filearg_ etc and then returns the path somehow).

PaulWessel commented:
> But files were not updated (the hash maybe). They have the same name and content. But note that all grids were deleted when only one was requested.

Yes, they have the same name, obviously.  If they downloaded then they must have new content since all the files on the server are new. There is no way to download the old file from the server.  If you do grdinfo on these files it will at the end say
earth_relief_01d.grd: scale_factor: 0.215942639865 add_offset: -1517 packed z-range: [-32764,32767]

The old files all had scale factor 1.

joa-quim commented:
I think you (or me) are not getting it all. I asked to download ``@earth_relief_01d``, which it did like requested but at same time it also deleted ALL the other *earth_reief_xxx* grids I had in that dir.

joa-quim commented:
The *gmt_check_filearg* case. It's not only MB who uses it. I do it too in my mbgmt module. And if we want people to write modules we must promote this function to an API one and not change it.

PaulWessel commented:
Too late for that function, but I am testing the one discussed above that updates the name with the path.  It may be a candideate for the API.  The old function will remain where it is for backwardness.

Yes, I understood all your earth_* files got deleted.  Next time you access any of them it will have to download it so you get the new data.  There is no option to say "I am OK with this set of files, do not update any further" unless you set the download size to zero or turn download off.

seisman commented:
The CI reports warnings:
```
gmt [WARNING]: File /home/vsts/.gmt/server/gmt_hash_server.txt said it has 226 records but only found 198 - download error???
gmt [WARNING]: File /home/vsts/.gmt/server/gmt_hash_server.txt will be deleted.  Please try again
```

PaulWessel commented:
Problem is that I placed the marles dirs in there.  I will move them out until we are ready and agree on names.  OK, now fixed, file lists 198 now.
